### PR TITLE
ENH: Add Series.dt.total_seconds GH #10817

### LINKED ISSUE #10817: Add .dt.total_seconds() method for timedelta64 Series
Background: I am trying to build a horizonal bar graph that shows me when various builds on a server started and ended, so that I can see visually which builds overlapped in time. But the idea that a horizontal bar graph's bar-width would be a timedelta seems completely foreign to all of the plotting libraries that I have tried today (matplotlib, bokeh, ggplot), so I want to punt and instead reduce my timedelta64's to a floating-point number of days. I can then fudge the axes to display dates anyway.

Problem: there is no easy way to move from a timedelta64 to the standard Python idea of an ordinal date. I would have to extract all of the components manually (days, seconds, fractions of a second) and then multiply them each by an appropriate factor and then add them back together.

The Python `timedelta` class has an alternative that I like very much: a `total_seconds()` method that expresses then entire value of the `timedelta` as a single scalar number of seconds.

This is not a perfect solution, since the factors it uses internally neglect the fact that a day might have included a leap second, but it works pretty well for non-astronomical use.

Could a `Series.dt.total_seconds()` method be added that returns the floating point value `days * 86400 + seconds + microseconds / 1e6`? It would be a great convenience for moving from date-land to scalar-land before plotting. Thanks!

### PR DESCRIPTION
**Pull request by sjdenny**
Implements a Series.dt.total_seconds method for timedelta64 Series.

closes #10817

### CODE REVIEW

**Comment by jreback** [comment]
does timedelta.total_seconds() provide fractional second as wel?

**Comment by sjdenny** [comment]
Yes, fractional seconds are included:

```
In [1]: s = Series(pd.timedelta_range('1day',periods=5,freq='1ms'))

In [2]: s
Out[2]: 
0          1 days 00:00:00
1   1 days 00:00:00.001000
2   1 days 00:00:00.002000
3   1 days 00:00:00.003000
4   1 days 00:00:00.004000
dtype: timedelta64[ns]

In [3]: s.dt.total_seconds
Out[3]: 
0    86400.000
1    86400.001
2    86400.002
3    86400.003
4    86400.004
dtype: float64
```

**Comment by jreback** [comment]
not what I mean

is the actual Python timedelta.total_seconds have fractions (I think yes) just confirming

**Comment by sjdenny** [comment]
Ah, I understand. Yes, it does:
```
In [38]: t = timedelta(days=1,microseconds=40)

In [39]: t.total_seconds()
Out[39]: 86400.00004
```

File: pandas/tseries/tests/test_timedeltas.py [hunk 1]
```diff
@@ -944,6 +945,27 @@ def test_fields(self):
 
         tm.assert_series_equal(s.dt.days,Series([1,np.nan],index=[0,1]))
         tm.assert_series_equal(s.dt.seconds,Series([10*3600+11*60+12,np.nan],index=[0,1]))
+        
+    def test_total_seconds(self):
```
Comment (jreback) [inline pandas/tseries/tests/test_timedeltas.py]: add a test in ``test_series/test_dt_namespace_accessor`` as well (its like the one you have for test Series)

File: pandas/tseries/tests/test_timedeltas.py [hunk 2]
```diff
@@ -944,6 +945,27 @@ def test_fields(self):
...
+        # test index
+        rng = timedelta_range('1 days, 10:11:12.100123456', periods=2, freq='s')
+        expt = [1*86400+10*3600+11*60+12+100123456./1e9,1*86400+10*3600+11*60+13+100123456./1e9]
+        assert_allclose(rng.total_seconds, expt, atol=1e-10, rtol=0)
+        
+        # test Series
+        s = Series(rng)
+        s_expt = Series(expt,index=[0,1])
+        tm.assert_series_equal(s.dt.total_seconds,s_expt)
```
Comment (jreback) [inline pandas/tseries/tests/test_timedeltas.py]: can you add a test for the scalar as well (even though we know it works as its a sub-class of ``timedelta``), I don't think its testsed. e.g. ``Timedelta(....).total_seconds()``.

Note that this is a method, so maybe we should make ``total_seconds`` (on a ``TimedeltaIndex``) a method as well (even though its a bit odd, for consitency with the scalar)?

**Comment by jorisvandenbossche** [comment]
Indeed, didn't see that, but +1 making it a method instead of a property for consistency with datetime.timedelta/pd.Timedelta

**Comment by jorisvandenbossche** [comment]
Further, @sjdenny can you add a whatsnew notice (see doc/source/whatsnew/v0.17.0.txt, somewhere in 'other enhancements')

**Comment by sjdenny** [comment]
In extending the (nanosecond-precision) tests, I've come across this behaviour:
```
In [18]: td = pd.Timedelta(nanoseconds=50)

In [19]: td.nanoseconds
Out[19]: 50

In [20]: td.total_seconds()
Out[20]: 0.0

In [21]: td.total_seconds() == 0.0
Out[21]: True

In [22]: td2 = pd.Timedelta(microseconds=50)

In [23]: td2.total_seconds()
Out[23]: 5e-05
```

It appears (e.g. [here](https://docs.python.org/2/library/datetime.html#datetime.timedelta.total_seconds)) that timedelta.total_seconds() aims for microsecond accuracy only. Is this the behaviour we want to reproduce in Series.dt.total_seconds()? Nanosecond precision appears preferable, but then the scalar and Series functions would have slightly different behaviour.

**Comment by jorisvandenbossche** [comment]
It should be consistent between both in any case, but maybe we can also see the `Timedelta` behaviour as a bug?

I suspect that this method is just inherited from datetime.timedelta, so maybe we will have to overwrite it ourselves in the subclass.

BTW, the reason of this difference is that datetime.timedelta does not support nanoseconds, while pd.Timedelta does

**Comment by jreback** [comment]
the pandas functions should work with ns precision in all cases

so need to override the scalar function as well

**Comment by jorisvandenbossche** [comment]
@sjdenny This should probably be added somewhere here: https://github.com/pydata/pandas/blob/master/pandas/tslib.pyx#L2415 (I think the `self.value` will give you the nanoseconds, and then you can use the same approach)

File: doc/source/whatsnew/v0.17.0.txt [hunk 3]
```diff
@@ -176,6 +176,10 @@ Other enhancements
 
 - ``pandas.tseries.offsets`` larger than the ``Day`` offset can now be used with with ``Series`` for addition/subtraction (:issue:`10699`).  See the :ref:`Documentation <timeseries.offsetseries>` for more details.
 
+- ``pd.Series`` of type timedelta64 has new method .dt.total_seconds() returning the duration of the timedelta in seconds (:issue: `10817`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.17.0.txt]: use double-backticks around ``timedelta64`` and ``.dt.total_seconds()``

File: pandas/tests/test_series.py [hunk 4]
```diff
@@ -142,6 +142,7 @@ def compare(s, name):
         for s in [Series(timedelta_range('1 day',periods=5),index=list('abcde')),
                   Series(timedelta_range('1 day 01:23:45',periods=5,freq='s')),
                   Series(timedelta_range('2 days 01:23:45.012345',periods=5,freq='ms'))]:
+            #assert False
```
Comment (jreback) [inline pandas/tests/test_series.py]: leftover?

File: pandas/tseries/tests/test_timedeltas.py [hunk 5]
```diff
@@ -945,6 +946,31 @@ def test_fields(self):
         tm.assert_series_equal(s.dt.days,Series([1,np.nan],index=[0,1]))
         tm.assert_series_equal(s.dt.seconds,Series([10*3600+11*60+12,np.nan],index=[0,1]))
 
+    def test_total_seconds(self):
+        # test index
```
Comment (jreback) [inline pandas/tseries/tests/test_timedeltas.py]: add the issue number as a comment here

**Comment by jreback** [comment]
I think also add the method ``.total_seconds()`` on the ``NaTType`` object (returns ``np.nan``) for compat. (and pls add a test as well).

File: pandas/tseries/tdi.py [hunk 6]
```diff
@@ -391,6 +391,17 @@ def f(x):
             result = result.astype('int64')
         return result
 
+    def total_seconds(self):
+        """ Total duration of each element expressed in seconds. """
+        values = self.asi8
+        hasnans = self.hasnans
+        result = 1e-9 * values
+        if hasnans:
```
Comment (jreback) [inline pandas/tseries/tdi.py]: use ``result = self._maybe_mask_results(result)`` (you can remove the ``hasnans`` as well)

**Comment by jreback** [comment]
@sjdenny awesome job!

**Comment by jorisvandenbossche** [comment]
@sjdenny Thanks a lot!

**Pull request merged**